File: src/Memories/Application/CreateMemory/CreateMemoryHandler.php
<?php

namespace App\Memories\Application\CreateMemory;

use App\Categories\Domain\CustomException;
use App\Memories\Domain\ForbiddenWords\ForbiddenWordChecker;
use App\Memories\Domain\Memory;
use App\Memories\Domain\MemoryCategories;
use App\Memories\Domain\MemoryContent;
use App\Memories\Domain\MemoryName;
use App\Memories\Domain\MemoryRepositoryInterface;
use App\Memories\Domain\MemoryType;
use App\Memories\Domain\SameTypeAndNameChecker;
use App\Memories\Domain\SameTypeAndNameException;
use App\Shared\Domain\Clock;
use App\Shared\Domain\EventBusInterface;
use App\Shared\Domain\UuidValueObject;

class CreateMemoryHandler
{
    public function __construct(
        private readonly MemoryRepositoryInterface $memoryRepository,
        private readonly Clock $clock,
        private readonly EventBusInterface $eventBus,
        private readonly SameTypeAndNameChecker $sameTypeAndNameChecker,
        private readonly ForbiddenWordChecker $forbiddenWordChecker,
    ) {
    }

    /**
     * @throws SameTypeAndNameException
     * @throws CustomException
     */
    public function execute(CreateMemoryCommand $command): void
    {
        $this->sameTypeAndNameChecker->assert($command->name, $command->type);

        $categories = new MemoryCategories();
        foreach ($command->categories as $category) {
            $categories->add(UuidValueObject::fromValue($category));
        }

        $memory = Memory::create(
            UuidValueObject::fromvalue($command->id),
            MemoryName::fromValue($command->name),
            MemoryType::fromValue($command->type),
            $this->clock->now(),
            $categories,
            UuidValueObject::fromValue($command->loggedUserId),
            $command->content ? MemoryContent::fromValue($command->content) : null,
            $this->forbiddenWordChecker,
        );

        $this->memoryRepository->save($memory);
        $this->eventBus->dispatchAll($memory->getEvents());
    }
}

File: src/main/kotlin/com/zaleort/second_brain/Memories/Application/UseCases/CreateMemory/CreateMemoryHandler.kt
package com.zaleort.second_brain.Memories.Application.UseCases.CreateMemory

import com.zaleort.second_brain.Memories.Application.UseCases.MemoryDTO
import com.zaleort.second_brain.Memories.Domain.Memory.Memory
import com.zaleort.second_brain.Memories.Domain.Memory.MemoryContent
import com.zaleort.second_brain.Memories.Domain.Memory.MemoryId
import com.zaleort.second_brain.Memories.Domain.Memory.MemoryTag
import com.zaleort.second_brain.Memories.Domain.Memory.MemoryTitle
import com.zaleort.second_brain.Memories.Domain.Memory.MemoryType
import com.zaleort.second_brain.Memories.Domain.Memory.MemoryRepositoryInterface
import com.zaleort.second_brain.Shared.Domain.Uuid.Uuid
import com.zaleort.second_brain.Tags.Domain.Tag.TagId
import com.zaleort.second_brain.Tags.Domain.Tag.TagRepositoryInterface
import com.zaleort.second_brain.Users.Domain.User.UserId
import org.springframework.stereotype.Service

@Service
class CreateMemoryHandler(
    val repository: MemoryRepositoryInterface,
    val tagRepository: TagRepositoryInterface,
) {
    fun execute(command: CreateMemoryCommand): MemoryDTO {
        val tags = command.tags.mapNotNull { tagRepository.findById(TagId(it)) }
            .map { MemoryTag(it.id.value, it.name.value, it.color?.value) }

        val memory = Memory.create(
            id = MemoryId(Uuid.random()),
            userId = UserId(command.userId),
            title = MemoryTitle(command.title),
            content = MemoryContent(command.content),
            type = MemoryType(command.type),
            tags = tags,
        )

        this.repository.save(memory)

        return MemoryDTO.fromMemory(memory)
    }
}